### PR TITLE
Add examples for tabular data with FT_Transformer in AutoMM 

### PR DESCRIPTION
**Pull request by yiqings**
*Issue #, if available:*

*Description of changes:*

**Motivations:** Examples of using FT_Transformer in AutoMM on 11 tabular datasets. Details of the datasets are described in Table 1 in [1], which covers (i) classification (both binary and multi-class) and (ii) regression tasks. 

**Availability to the datasets**: The [official implementation](https://github.com/Yura52/tabular-dl-revisiting-models) provides the datasets in `numpy.darray` format: https://www.dropbox.com/s/o53umyg6mn3zhxy/data.tar.gz?dl=1. Data in `numpy.darray` formatting was first pre-processedin into `.csv` format, which can thus be loaded with `pandas.Dataframe` as the input to AutoMMPredictor.  All Data in `.csv` format can be downloaded from s3.

**Availability to the results** The trained models, configs can be downloaded from https://autogluon.s3.us-west-2.amazonaws.com/results/tabular/results.zip.

**ToDo:**
1. ~~Refine the dataset loading (from numpy.darray to pandas.Dataframe).~~
2. ~~Add hyperparameters.~~ 
3. ~~Add README file.~~

**Results**
Datasets | ca | ad | he | ja | hi | al | ep | ye | co | ya | ml 
 ----  | ----  | ----  | ----  | ----  | ----  | ----  | ----  | ----  | ----  | ----  | ----  
 metrics | rmse | acc | acc | acc | acc | acc | acc | rmse | acc | rmse | rmse
 problem_type | regression | binary | multiclass | multiclass | binary | multiclass | binary | regression | multiclass | regression | regression
 #objects | 20640 | 48842 | 65196 | 83733 | 98050 | 108000 | 500000 | 515345 | 581012 | 709877 | 1200192
 #num. | 8 | 6 | 27 | 54 | 28 | 128 | 2000 | 90 | 54 | 699 | 136
 #cat. | 0 | 8 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0
 #classes | - | 2 | 100 | 4 | 2 | 1000 | 2 | - | 7 | - | -
 Best in [1] | 0.459 | 0.859 | 0.396 | 0.732 | 0.729 | 0.963 | 0.8982 | 8.794 | 0.970 | 0.753 | 0.745
 FT-Transformer in [1] | 0.459 | 0.859 | 0.391 | 0.732 | 0.729 | 0.960 | 0.8982 | 8.855 | 0.970 | 0.756 | 0.746
 AutoMM FT-Transformer + Const Lr=1e-4 | 0.404 | 0.863 | 0.388 | 0.727 | 0.729 | 0.954 | OverflowError | 0.781 | 0.964 | AttributeError | 0.904
 AutoMM FT-Transformer + Cosine Lr=1e-4| 0.401 | 0.864 | 0.388 | 0.728 | 0.729 | 0.952 | OverflowError | 0.780 | 0.964 | 0.765 | 0.924

**Reference:**
[1] [Revisiting Deep Learning Models for Tabular Data](https://proceedings.neurips.cc/paper/2021/file/9d86d83f925f2149e9edb0ac3b49229c-Paper.pdf) (NeurIPS 2021)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-1752-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1752/1/index.html

**Comment by sxjscience** [comment]
Need to add a `README.md`.

**Comment by sxjscience** [comment]
@zhiqiangdon We need to take a look at the overflow error.

File: examples/automm/utils.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        METRIC: RMSE,
+    },
+    'ye': {
+        FULLNAME: 'year',
+        TYPE: REGRESSION,
+        METRIC: RMSE,
+    },
+}
+
+def load_data_per_split(
```
Comment (Innixma) [inline examples/automm/utils.py:76]: Why is this so complicated? Why aren't the datasets in parquet format? How do I access these datasets?

File: examples/automm/example_tabular.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        label='0', 
+        problem_type=problem_type,
+        eval_metric=metric,
+        path=args.exp_dir,
+        verbosity=4,
+    )
+
+    predictor.fit(
+        train_data=train_df,
+        tuning_data=test_df,
```
Comment (Innixma) [inline examples/automm/example_tabular.py:22]: Why are we passing test data as tuning data????
Why are we never using val_df?
How are you calculating scores for a dataset, that isn't part of this code?

File: examples/automm/example_tabular.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        column_types=column_types,
```
Comment (Innixma) [inline examples/automm/example_tabular.py:23]: why are we passing column_types? Isn't the dataframe dtypes enough?

File: examples/automm/example_tabular.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--dataset_dir', default='./data', type=str)
+    parser.add_argument('--dataset_name', default='ad', type=str)
+    parser.add_argument('--exp_dir', default='./result/', type=str)
+    parser.add_argument('--seed', default=95, type=int)
+
+    args = parser.parse_args()
+
+    main(args)
```
Comment (Innixma) [inline examples/automm/example_tabular.py:36]: missing newline at end of file

File: examples/automm/example_tabular.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        column_types=column_types,
+        seed=args.seed,
+    )
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--dataset_dir', default='./data', type=str)
+    parser.add_argument('--dataset_name', default='ad', type=str)
+    parser.add_argument('--exp_dir', default='./result/', type=str)
+    parser.add_argument('--seed', default=95, type=int)
```
Comment (Innixma) [inline examples/automm/example_tabular.py:32]: why seed 95? Why not just seed=0?

File: examples/automm/example_tabular.py (hunk 2 above)
Comment (sxjscience) [reply examples/automm/example_tabular.py:22]: I think it should be `val_df` @yiqings

File: examples/automm/dataset.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    def metric(self):
+        return ACC
+    
+    @property
+    def problem_type(self):
+        return BINARY
+    
+
+adultdataset = AdultTabularDataset('test','/home/ubuntu/dataset')
+print(adultdataset.data)
```
Comment (sxjscience) [inline examples/automm/dataset.py]: nit: Add new line here

**Comment by szha** [comment]
Job PR-1752-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1752/4/index.html

**Comment by szha** [comment]
Job PR-1752-5 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1752/5/index.html

File: examples/automm/example_tabular.py (hunk 3 above)
Comment (yiqings) [reply examples/automm/example_tabular.py]: Although `column_types` is optional, we found that very occasionally the column_types may be inferred errorly (for example some NUMERICAL columns are identified as CATEGORY). 
In our currently implementation, the model configurations for [NUMERICAL] and [CATEGORY,NUMERICAL] are different:
- [NUMERICAL]: model.numerical_transformer.num_trans_blocks = 3;
- [CATEGORY,NUMERICAL]: model.numerical_transformer.num_trans_blocks = 0, model.categorical_transformer.num_trans_blocks = 0, and model.fusion_transformer.num_trans_blocks = 3.
Hence, passing column_types is necessary.

We'll update `create_model` to make sure that fusion_transformer is applied to [CATEGORY,NUMERICAL], [NUMERICAL] and [CATEGORICAL]. Afterwards, passing `column_types` will not be necessary.

File: examples/automm/example_tabular.py (hunk 5 above)
Comment (yiqings) [reply examples/automm/example_tabular.py]: We set the default to 0 now.

File: examples/automm/utils.py (hunk 1 above)
Comment (yiqings) [reply examples/automm/utils.py]: Yes. We refine the dataset loading thoroughly. We now support downloading datasets from s3. In README.md, we have one section to describe the accessibility to the data.

File: examples/automm/example_tabular.py (hunk 2 above)
Comment (yiqings) [reply examples/automm/example_tabular.py]: Yes. Updated.

File: examples/automm/dataset.py [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+    def label_type(self):
+        return NUMERICAL
+    
+    @property
+    def metric(self):
+        return RMSE
+    
+    @property
+    def problem_type(self):
+        return REGRESSION
```
Comment (sxjscience) [inline examples/automm/dataset.py:490]: We may refactor this part later. We can release the auto_mm_bench package or reuse the huggingface datasets API.

File: examples/automm/utils.py [hunk 8]
```diff
@@ -0,0 +1,237 @@
+'''
```
Comment (sxjscience) [inline examples/automm/utils.py:1]: I think it makes sense to move the download utils to https://github.com/awslabs/autogluon/blob/master/text/src/autogluon/text/automm/utils.py

**Comment by szha** [comment]
Job PR-1752-6 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1752/6/index.html

**Review by sxjscience — APPROVED** [review]
LGTM

**Comment by szha** [comment]
Job PR-1752-10 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1752/10/index.html

File: examples/automm/README.md [hunk 9]
```diff
@@ -0,0 +1,74 @@
+Examples showing how to use AutoMMPredictor.
+
+
+## 1. Tabular Data
+
+### 1.1 Example
+[`example_tabular.py`](./example_tabular.py) : This example provide a use case for the pure *tabular* data, including numerical and categorical, with FT_Transformer [1].
```
Comment (zhiqiangdon) [inline examples/automm/README.md:7]: provide -> provides

**Review by zhiqiangdon — APPROVED** [review]
Nice scripts and examples!

**Pull request merged**